### ISSUE TITLE
Sequential approach too slow for big image sets

### ISSUE THREAD

**Issue by Expro** [chaiNNer-org/chaiNNer#1274]
Currently, chaiNNer does everything sequentially. If iterating over directory full of images, main part of the loop is:

Load Image -> Upscale Image -> Save Image

Issue is: every step is performed synchronously. It means that in "Load Image" step GPU is idling, in "Upscale Image" CPU and disk is idling, and in "Save Image" GPU is idling again. As a result, chaiNNer is not able to fully utilize available resources and is just slow. For example, I have such loop running now and GPU is loaded at 10%, SSD is at 3% and CPU is at 40% (and only because that's old, small CPU). Hardware is bored and underutilized, it's the chaiNNer that is not able to use it properly.

Entire loop should be moving with speed of slowest part of hardware - one of this (CPU, GPU, disk) should be utilized at 100%, most likely GPU.

Realistically speaking, each of this steps should have it's own queue and operations should be performed asynchronously. While GPU is upscaling image, next image should be already loaded or be loading. There is also no point in waiting on image to be saved to disk before moving to next step - just schedule it as asynchronous operation and move on.

That also touches a bit on my other issue - multi GPU support. chaiNNer already has phenomenal UI / UX and does wonderful things for ease of use, but is not yet architected for performance and thus slow to the point of being impractical.

**Comment by joeyballentine** [comment]
I completely agree. I just have no idea how to actually do that. It's possible @RunDevelopment would be able to handle that at some point though, maybe. But I know he has other stuff he wants to get done in the near future first. You wouldn't happen to know a lot about python multithreading would you?

**Comment by Expro** [comment]
> You wouldn't happen to know a lot about python multithreading would you?

I actually do know a lot about thread and task parallelism... in C#. Maybe I will take a crack at it, but I would have to learn Python on the fly. But it would definitely be better if someone fluent do that instead.
